sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,472 @@
+      mc_momentum = particle->P();
+      mc_pXEnd = particle->EndPx();
+      mc_pYEnd = particle->EndPy();
+      mc_pZEnd = particle->EndPz();
+      mc_energyEnd = particle->EndE();
+      mc_mass = particle->Mass();
+      mc_theta_xy = TMath::RadToDeg() * TMath::ATan(mc_pX0/mc_pY0);
+      mc_theta_yz = TMath::RadToDeg() * TMath::ATan(mc_pY0/mc_pZ0);
+      mc_theta_xz = TMath::RadToDeg() * TMath::ATan(mc_pX0/mc_pZ0);
+      mc_length = TrueTrackLength(particle);
+
+      if(n_tracks_map[mc_trackID] > 0 || n_showers_map[mc_trackID] > 0) reco_isReconstructed = true;
+      reco_nTracks = n_tracks_map[mc_trackID];
+      reco_nShowers = n_showers_map[mc_trackID];
+      reco_shower_purity = shower_pur_map[mc_trackID];
+      reco_shower_completeness = shower_comp_map[mc_trackID];
+      reco_track_purity = track_pur_map[mc_trackID];
+      reco_track_completeness = track_comp_map[mc_trackID];
+      reco_track_length = track_length_map[mc_trackID];
+      reco_shower_dEdx = shower_dEdx_map[mc_trackID];
+
+      fParticleTree->Fill();
+    }
+  }
+}
+
+std::vector<art::Ptr<recob::PFParticle> > RecoEff::GetPrimaryPFPs(art::Event const &e)
+{
+  art::Handle<std::vector<recob::PFParticle> > handlePFPs;
+  e.getByLabel(fPFParticleModuleLabel,handlePFPs);
+
+  std::vector<art::Ptr<recob::PFParticle> > primaries;
+
+  for(unsigned int pfp_i = 0; pfp_i < handlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(handlePFPs,pfp_i);
+    if(pfp->IsPrimary()){
+      if(std::abs(pfp->PdgCode()) == 12 || std::abs(pfp->PdgCode()) == 14) {
+        primaries.push_back(pfp);
+      }
+    }
+  }
+  return primaries;
+}
+
+art::Ptr<recob::PFParticle> RecoEff::GetPFP(art::Event const &e, long unsigned int const &id)
+{
+  art::Handle<std::vector<recob::PFParticle> > handlePFPs;
+  e.getByLabel(fPFParticleModuleLabel,handlePFPs);
+
+  const art::Ptr<recob::PFParticle> nullReturn;
+
+  for(unsigned int pfp_i = 0; pfp_i < handlePFPs->size(); ++pfp_i) {
+    const art::Ptr<recob::PFParticle> pfp(handlePFPs,pfp_i);
+    if(pfp->Self() == id) return pfp;
+  }
+  return nullReturn;
+}
+
+float RecoEff::Purity(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return (objectHits.size() == 0) ? def_float : objectHitsMap[trackID]/static_cast<float>(objectHits.size());
+}
+
+float RecoEff::Completeness(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID)
+{
+  std::map<int,int> objectHitsMap;
+
+  for(unsigned int i = 0; i < objectHits.size(); ++i) {
+    objectHitsMap[TruthMatchUtils::TrueParticleID(clockData,objectHits[i],true)]++;
+  }
+  return (hits_map[trackID] == 0) ? def_float : objectHitsMap[trackID]/static_cast<float>(hits_map[trackID]);
+}
+
+float RecoEff::TrackLength(art::Ptr<recob::Track> const &track)
```

Comment:
can this whole function be replaced with just `track.Length()`? If not, what is the difference?

Reply:
Its a very fair point! They only seem to differ to `~10^5` so it seems I was somewhat reinventing the wheel. I've taken it out :)